### PR TITLE
Int1e 1st derivative

### PR DESCRIPTION
**Pull request by henryw7**
Average speedup compared to fakemol implementation is 50%

### CODE REVIEW

File: gpu4pyscf/lib/gint/g3c1e_ip.cu [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+                        }
+                    }
+                }
+            }
+        }
+        deri_dCx_pair_sum += deri_dCx_per_pair;
+        deri_dCy_pair_sum += deri_dCy_per_pair;
+        deri_dCz_pair_sum += deri_dCz_per_pair;
+    }
+    atomicAdd(output + task_grid + ngrids * 0, deri_dCx_pair_sum);
```
Comment (wxj6000) [inline gpu4pyscf/lib/gint/g3c1e_ip.cu:336]: Reduce across  ao pair direction before atomicadd operation.

File: gpu4pyscf/lib/gint/g3c1e_ip.cu [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+    const int j0 = ao_loc[jsh] - ao_offsets_j;
+    const int n_density_elements_i = (i_l + 1) * (i_l + 2) / 2;
+    const int n_density_elements_j = (j_l + 1) * (j_l + 2) / 2;
+    const int n_density_elements_ij = n_density_elements_i * n_density_elements_j;
+    for (int j = 0; j < n_density_elements_j; j++) {
+        for (int i = 0; i < n_density_elements_i; i++) {
+            const double deri_dAx = output_cache[i + j * n_density_elements_i + 0 * n_density_elements_ij];
+            const double deri_dAy = output_cache[i + j * n_density_elements_i + 1 * n_density_elements_ij];
+            const double deri_dAz = output_cache[i + j * n_density_elements_i + 2 * n_density_elements_ij];
+            atomicAdd(output + ((i + i0) + (j + j0) * stride_j + 0 * stride_ij), deri_dAx);
```
Comment (wxj6000) [inline gpu4pyscf/lib/gint/g3c1e_ip.cu:245]: Reduce across grids direction before atomicadd.

File: gpu4pyscf/lib/gint/g1e_ip_root_1.cu [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+                deri_dCz_per_pair += prefactor * R000_1 * PCz;
+            }
+        }
+
+        const double D = density[bas_ij - hermite_density_offsets.pair_offset_of_angular_pair + hermite_density_offsets.density_offset_of_angular_pair];
+        deri_dCx_pair_sum += deri_dCx_per_pair * D;
+        deri_dCy_pair_sum += deri_dCy_per_pair * D;
+        deri_dCz_pair_sum += deri_dCz_per_pair * D;
+    }
+    atomicAdd(output + task_grid + 0 * ngrids, deri_dCx_pair_sum);
```
Comment (wxj6000) [inline gpu4pyscf/lib/gint/g1e_ip_root_1.cu:281]: Block reduction in ao pair.

File: gpu4pyscf/lib/gint/g1e_ip_root_1.cu [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+        const double charge = grid_point[3];
+        deri_dAx_grid_sum += deri_dAx_per_grid * charge;
+        deri_dAy_grid_sum += deri_dAy_per_grid * charge;
+        deri_dAz_grid_sum += deri_dAz_per_grid * charge;
+    }
+
+    const int* ao_loc = c_bpcache.ao_loc;
+    const int i0 = ao_loc[ish] - ao_offsets_i;
+    const int j0 = ao_loc[jsh] - ao_offsets_j;
+    atomicAdd(output + (i0 + j0 * stride_j + 0 * stride_ij), deri_dAx_grid_sum);
```
Comment (wxj6000) [inline gpu4pyscf/lib/gint/g1e_ip_root_1.cu:208]: Block reduction in grids.

File: gpu4pyscf/gto/int3c1e.py [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+
+def get_int3c1e_density_contracted(mol, grids, charge_exponents, dm, intopt):
+    omega = mol.omega
+    assert omega >= 0.0, "Short-range one electron integrals with GPU acceleration is not implemented."
+
+    dm = cp.asarray(dm)
+    if dm.ndim == 3:
+        if dm.shape[0] > 2:
+            print("Warning: There are more than two density matrices to contract with one electron integrals, "
+                  "it's not from an unrestricted calculation, and we're unsure about your purpose. "
```
Comment (wxj6000) [inline gpu4pyscf/gto/int3c1e.py:374]: "Warning: more than two density matrices are found for int3c1e kernel. They will be summed up to one density matrix."

File: gpu4pyscf/lib/gint/g3c1e_ip.cu (hunk 2 above)
Comment (henryw7) [reply gpu4pyscf/lib/gint/g3c1e_ip.cu:362]: Block reduction shows no performance improvement, so I prefer to keep it in a simpler atomicadd form.

**Review by wxj6000 — APPROVED** [review]

**Pull request merged**